### ISSUE TITLE
ICMP DNS lookup inaccuracy 

### ISSUE THREAD

**Issue by 2color** [prometheus/blackbox_exporter#208]
## Problem

It seems that the result coming from the blackbox exporter is inaccurate when using a DNS record.

When doing an ICMP probe of the same target in two ways:
- With the DNS record:
```
# HELP probe_dns_lookup_time_seconds Returns the time taken for probe dns lookup in seconds
# TYPE probe_dns_lookup_time_seconds gauge
probe_dns_lookup_time_seconds 6.3e-08
# HELP probe_duration_seconds Returns how long the probe took to complete in seconds
# TYPE probe_duration_seconds gauge
probe_duration_seconds 0.022335252
# HELP probe_ip_protocol Specifies whether probe ip protocol is IP4 or IP6
# TYPE probe_ip_protocol gauge
probe_ip_protocol 4
# HELP probe_success Displays whether or not the probe was a success
# TYPE probe_success gauge
probe_success 1
```
- Directly passing the IP
```
# HELP probe_dns_lookup_time_seconds Returns the time taken for probe dns lookup in seconds
# TYPE probe_dns_lookup_time_seconds gauge
probe_dns_lookup_time_seconds 2.24e-07
# HELP probe_duration_seconds Returns how long the probe took to complete in seconds
# TYPE probe_duration_seconds gauge
probe_duration_seconds 0.001478916
# HELP probe_ip_protocol Specifies whether probe ip protocol is IP4 or IP6
# TYPE probe_ip_protocol gauge
probe_ip_protocol 4
# HELP probe_success Displays whether or not the probe was a success
# TYPE probe_success gauge
probe_success 1
```

The difference is big and the DNS lookup time is too small to explain the difference in time.

### Host operating system: output of `uname -a`

`Linux 4.9.24-coreos #1 SMP Tue Apr 25 02:06:28 UTC 2017 x86_64 Intel(R) Xeon(R) CPU E5-2620 0 @ 2.00GHz GenuineIntel GNU/Linux`

### blackbox_exporter version: output of `blackbox_exporter -version`

0.8.1

### What is the blackbox.yml module config.
```
modules:
  icmp:
    prober: icmp
    timeout: 5s
    icmp:
      preferred_ip_protocol: "ip4"
```

### What is the prometheus.yml scrape config.

```
  - job_name: 'blackbox'
    metrics_path: /probe
    params:
      module: [icmp]
    static_configs:
      - targets:
        - ping-test-db.SOME_URL.aivencloud.com
    relabel_configs:
      - source_labels: [__address__]
        target_label: __param_target
      - source_labels: [__param_target]
        target_label: instance
      - target_label: __address__
        replacement: localhost:9115  # Blackbox exporter.
```

### What logging output did you get from the `blackbox_exporter`?

See above

### What did you expect to see?

I expected the `probe_dns_lookup_time_seconds` when probing the DNS record `ping-test-db.SOME_URL.aivencloud.com` to be larger to explain the value difference in the `probe_duration_seconds` metric.

**Comment by brian-brazil** [comment]
What does tcpdump say is happening?

**Comment by 2color** [comment]
> What does tcpdump say is happening?

Where should I run tcpdump? on the host machine of the container? or the container?

**Comment by brian-brazil** [comment]
Host machine, and I'd recommend getting anything like containers out of the picture to make debugging easier.

**Comment by songcaixin** [comment]
Have you solved this problem, please?When doing an probe, the DNS lookup time is also too small ,but i don't know why....

**Comment by SuperQ** [comment]
This feels like it could be caching related. What is the local resolver config?

**Comment by songcaixin** [comment]
It's as follows:
# Dynamic resolv.conf(5) file for glibc resolver(3) generated by resolvconf(8)
#     DO NOT EDIT THIS FILE BY HAND -- YOUR CHANGES WILL BE OVERWRITTEN
# 127.0.0.53 is the systemd-resolved stub resolver.
# run "systemd-resolve --status" to see details about the actual nameservers.

nameserver 127.0.0.53

**Comment by songcaixin** [comment]
It is as follows:
nameserver 127.0.0.53

**Comment by SuperQ** [comment]
That's what I figured, you have a local resolver that positively caches the A record, but if the IP lookup has no PTR, it will get an `NXDOMAIN` and not cache that.

**Comment by frittentheke** [comment]
@brian-brazil I just ran into the exact same issue. I tcpdumped the DNS lookup and it's blazing fast and no query timing out or getting any NXDOMAIN (of that would confuse the blackbox exporter).

TL;DR: If you use a host name instead of just an IP there is something seriously wrong with the probe_duration metric.

**Comment by brian-brazil** [comment]
Try the latest version, there was a bug fixed in this area.

**Comment by frittentheke** [comment]
Ah sorry, I just updated my post mentioning that I am using 0.11.0

**Comment by brian-brazil** [comment]
If you believe there's a bug please file a new issue, including full debug output.
